File: 프로그래머스/Level2/방문길이.py
# https://programmers.co.kr/learn/courses/30/lessons/49994

def solution(dirs):
    history = dict()
    now = (0, 0)
    for dir in dirs:
        x, y = now
        if dir == 'U':
            if y + 1 <= 5:
                now = (x, y + 1)
                if (x,y,x,y+1) not in history:
                    history[(x,y,x,y+1)] = True
        elif dir == 'D':
            if y - 1 >= -5:
                now = (x, y - 1)
                if (x,y-1,x,y) not in history:
                    history[(x,y-1,x,y)] = True
        elif dir == 'R':
            if x + 1 <= 5:
                now = (x + 1, y)
                if (x,y,x+1,y) not in history:
                    history[(x,y,x+1,y)] = True
        elif dir == 'L':
            if x - 1 >= -5:
                now = (x - 1, y)
                if (x-1,y,x,y) not in history:
                    history[(x-1,y,x,y)] = True
    return len(history)

File: 프로그래머스/Level2/방문길이.js
// https://programmers.co.kr/learn/courses/30/lessons/49994

const solution = (dirs) => {
	const history = {};
	let nowX = 0,
		nowY = 0;
	dirs.split('').forEach((d) => {
		switch (d) {
			case 'U':
				if (nowY < 5) {
					const nextX = nowX;
					const nextY = nowY + 1;
					const temp = [nowX, nowY, nextX, nextY];
					if (temp in history === false) history[temp] = 1;
					nowY++;
				}
				break;
			case 'D':
				if (nowY > -5) {
					const nextX = nowX;
					const nextY = nowY - 1;
					const temp = [nextX, nextY, nowX, nowY];
					if (temp in history === false) history[temp] = 1;
					nowY--;
				}
				break;
			case 'R':
				if (nowX < 5) {
					const nextX = nowX + 1;
					const nextY = nowY;
					const temp = [nowX, nowY, nextX, nextY];
					if (temp in history === false) history[temp] = 1;
					nowX++;
				}
				break;
			case 'L':
				if (nowX > -5) {
					const nextX = nowX - 1;
					const nextY = nowY;
					const temp = [nextX, nextY, nowX, nowY];
					if (temp in history === false) history[temp] = 1;
					nowX--;
				}
				break;
		}
	});
	return Object.keys(history).length;
};
